Improve tagging for a (maybe) null error

jdleesmiller opened the pull request:
When trying to add some error tagging, I noticed that fairly often a callback is just passing a possible error up the chain, e.g.
```js
function foo(callback) {
  bar((err, ignoredResult) => callback(err))
}
// or just
function foo(callback) {
  bar((err) => callback(err))
}
// or, if we are OK with passing back the ignored result, just
function foo(callback) {
  bar(callback)
}
```

At the moment, tagging the error in those cases is quite cumbersome, because it might be (probably will be) null, and `OError.tag` can't handle a null. So the one-liner has to turn into a full error check:
```js
function foo(callback) {
  bar((err, ignoredResult) => {
    if (err) return callback(OError.tag(err))
    callback()
   }
}
```

If `OError.tag` is a no-op when the error is null, then we can write
```js
function foo(callback) {
  bar((err) => callback(OError.tag(err)))
}
```
instead, which is a bit shorter.

I wondered about instead / also providing some sort of callback wrapper, along the lines of
```js
OError.tagErrback = (callback, message, info) => {
  return (err, ...args) => {
    if (err) {
      callback(OError.tag(err, message, info))
    } else {
      callback(null, ...args)
    }
  }
}

// usage:
function foo(callback) {
  bar(OError.tagErrback(callback))
}
// or
function foo(callback) {
  bar(OError.tagErrback(callback, 'failed to bar'))
}
```
but that seems a bit magical and hard to name, and we're trying to move toward async/await anyway, so I thought maybe this simple change to ignore a null `err` would suffice.

I added two more unrelated examples while updating the docs for this case (they get copied from the jsdoc to the readme and the .d.ts).

This could go into a minor release.

das7pad commented on index.js:
```diff
@@ -44,12 +44,36 @@ class OError extends Error {
+   * @param {Error | null | undefined} error the error to tag (no-op if missing)
    * @param {string} [message] message with which to tag `error`
    * @param {Object} [info] extra data with wich to tag `error`
    * @return {Error} the modified `error` argument
```
This is not accurate anymore.

jdleesmiller replied:
Good point, thanks!

das7pad reviewed (commented):
I think the alternative `OError.tagErrback` approach does not generate a meaningful traceback.

In order to preserve the (possibly enforced) jsdoc type annotation for the error param (and return value usage) in regular tagging, WDYT about:

```js 
OError.tagIfExists(error, message, info) {
  if (!error) return undefined
  return OError.tag(error, message, info)
}
```
This provides an opt-in method for the 'magic'/simplified tagging (likely used in callbacks only).

mserranom commented:
Looks good. I prefer ignoring `error` over an `OError.tagErrback` or `OError.tagIfExists`. Invoking `callback(err)` is pretty idiomatic in javascript, and sounds simpler to offer a wrapper that can be used with the same expectations.

emcsween reviewed (approved):
That seems reasonable. I'm fine with either the current version or `tagIfExists()`.

jdleesmiller commented:
Thanks, all.

- I was annoyed that the typecheck still passed despite my error (that @das7pad spotted), so I decided to get it checked with `--strict`. Predictably, this resulted in a minor odyssey, but the required fixes turned out to be relatively simple in the end. 3af5ae3
   - I think the type of `withInfo` was also wrong --- I don't think there's a use case for it to be undefined. Technically this is a breaking change, but I don't think it is likely to cause any problems in actual code.
   - It found an invalid expectation (`to.be.truthy`).

- I went with `tagIfExists` to preserve the existing stricter signature for `tag`. (This undoes some of the previous commit.) 16bc1c2

   - The problem with the simple approach of calling `tag` from `tagIfExists` is that `tag` then turns up in the stacktrace. I extracted a private `_tag` helper to avoid this, at least when we can use `Error.captureStackTrace`. If we can't use it, the stack trace will be marginally worse with this change; one solution would be to just copy the body of the tag method, but that seemed untidy.

- Then coveralls complained that the coverage decreased, so I tested a couple of extra edge cases. I guess that is a good thing. d42baba

jdleesmiller commented:
Actually, maybe hold those reviews... now that this is more complicated, I think maybe I should just do
```js
function foo(callback) {
  bar((err) => callback(err && OError.tag(err)))
}
```
That seems to work, is still one line, is not too hard to parse (I think), and doesn't complicate the library.

I'll document that and keep the strict mode and some of the other cleanup in this PR.

And coveralls remains unhappy, it seems.

das7pad commented on test/o-error-util.test.js:
```diff
@@ -149,9 +149,44 @@ describe('OError.tag', function () {
   })
+
+  it('is not included in the stack trace if using capture', function () {
+    if (!Error.captureStackTrace) return
```
```suggestion
    if (!Error.captureStackTrace) return this.skip()
```
Let's flag that we are not evaluating this test case when `Error.captureStackTrace` is missing. [Mocha docs are here.](https://mochajs.org/#inclusive-tests)

jdleesmiller commented:
OK, I've removed the `tagIfExists` and hopefully kept the other improvements. Maybe @das7pad can now give this a read?

I added some more tests to satisfy coveralls.

I'll probably do a squash merge.

das7pad commented on index.js:
```diff
@@ -106,7 +127,7 @@ class OError extends Error {
     const oError = /** @type{OError} */ (error)
 
-    let stack = oError.stack
+    let stack = oError.stack || '(no stack)'
```
Just noting that this is now possibly generating

```
SomeTopLevelError: msg
...
cause by:
   (no stack)
```

It was _broken_ before, so I'd leave it as is (proposed in this PR).

<details>

```js
> const OError = require('.')
> try { throw 'msg-lib' } catch (errLib) { err = new OError('msg-app').withCause(errLib); console.log({errLib, stack: errLib.stack, fullStack: OError.getFullStack(err)}) }
{
  errLib: 'msg-lib',
  stack: undefined,
  fullStack: 'OError: msg-app\n' +
    '    at repl:1:48\n' +
    '    at Script.runInThisContext (vm.js:120:20)\n' +
    '    at REPLServer.defaultEval (repl.js:430:29)\n' +
    '    at bound (domain.js:426:14)\n' +
    '    at REPLServer.runBound [as eval] (domain.js:439:12)\n' +
    '    at REPLServer.onLine (repl.js:758:10)\n' +
    '    at REPLServer.emit (events.js:323:22)\n' +
    '    at REPLServer.EventEmitter.emit (domain.js:482:12)\n' +
    '    at REPLServer.Interface._onLine (readline.js:322:10)\n' +
    '    at REPLServer.Interface._line (readline.js:699:8)\n' +
    'caused by:\n' +
    '    (no stack)'
}
```
</details>
